kano_updater/monitor.py

```diff
@@ -0,0 +1,165 @@
+            coll = defaultdict(set)
+
+            for line in stdout.split('\n'):
+                items = line.split()
+                if len(items) >= 2:
+                    ppid, pid = items[:2]
+                    if ppid == 'PPID':
+                        continue
+            try:
+                coll[int(ppid)].add(int(pid))
+            except:
+                pass  # ignore int conversion failure
+
+            def closure(pids):
+                curr = pids.copy()
+                for p in pids:
+                    curr = curr.union(closure(coll[p]))
+                return curr
+            return closure(set([self.top_pid]))
+
+        except subprocess.CalledProcessError:
+            return set()
+
+    def isChanged(self):
+        """
+        return True if we believe it is making progress
+        otherwise False
+        """
+        changed = False
+        new_children = self._get_children()
+
+        if self.curr_children != new_children:
+            self.curr_children = new_children
+            changed = True
+        return changed
+
+
+def monitor(watchproc, timeout):
+    """
+    Monitor a process tree. If the (recursive) set of child processes changes
+    or we are sent a SIGUSR1, we note that it is making progress.
+    if it has not made progress for `timeout` seconds, or the process
+    finished, exit.
+
+    Returns true if we timed out and false if the process finished
+
+
+    """
+    watchpid = watchproc.pid
+
+    spoll = kano_updater.utils.signalPoll(signal.SIGUSR1)
+
+    lastEvent = time.time()
+
+    pids = monitorPids(watchpid)
+
+    while True:
+        now = time.time()
+        # check for child events
+        changed = pids.isChanged()
+        if watchproc.poll() is not None:
+            return False
+
+        signalled = spoll.poll()
+        if changed or signalled:
+            lastEvent = now
+
+        if lastEvent + timeout < now:
+            return True
+
+        time.sleep(1)
+
+
+def heartbeat():
+    """
+    Inform monitor process, if it exists, that we are still alive
+    """
+
+    monitor_pid = os.environ.get("MONITOR_PID")
+    if monitor_pid:
+        try:
+            pid = int(monitor_pid)
+            os.kill(pid, signal.SIGUSR1)
+        except:
+            logger.error("Invalid monitor pid {}".format(monitor_pid))
+
+
+def run(cmdargs):
+    os.environ["MONITOR_PID"] = str(os.getpid())
+    subproc = subprocess.Popen(cmdargs,
+                               shell=False)
+
+    if not monitor(subproc, MONITOR_TIMEOUT):
+        return subproc.returncode
+    else:
+        if '--gui' in cmdargs:
+            from kano.gtk3 import kano_dialog
+            kdialog = kano_dialog.KanoDialog(
+                _("Update error"),
+                _("The updater seems to have got stuck. Press OK to reboot"))
+            kdialog.run()
+            os.system('systemctl reboot')
+
+        else:
+            return TIMEOUT_RC
+
+
+def manual_test_main(argv):
```

Comment:
Maybe instead add a directory into the `tests` directory for manual tests and put this there?